asteroid/engine/solver.py

```diff
@@ -95,69 +110,92 @@ def train(self):
-            else:
-                est_targets = self.model(inputs)
-            loss = self.loss_class.compute(targets, est_targets, infos=infos)
-
-            if not validation and not torch.isnan(loss):
-                self.optimizer.zero_grad()
-                loss.backward()
-                # if isinstance(self.model.module.encoder, AnalyticFreeFB):
-                #     value_if_nan(self.model.module.encoder.filters.grad)
-                #     value_if_nan(self.model.module.decoder.filters.grad)
-                torch.nn.utils.clip_grad_norm_(self.model.parameters(),
-                                               self.max_norm)
-                self.optimizer.step()
-
+            loss = self.run_one_batch(data, validation=validation)
             total_loss += loss.item()
-
-            if i % self.print_freq == 0:
+            if i % self.tr_conf['print_freq'] == 0:
                 print('Epoch {0} | Iter {1} | Average Loss {2:.3f} | Current'
                       ' Loss {3:.6f} | {4:.1f} ms/batch '
                       ''.format(epoch+1, i+1, total_loss/(i+1), loss.item(),
                                 1000*(time.time()-start)/(i+1)), flush=True)
         return total_loss / (i + 1)
 
+    def run_one_batch(self, data, validation=False):
+        """ Run forward backward step for one batch."""
+        inputs, targets, infos = self.unpack_data(data)
+        est_targets = self.forward_model(inputs, validation=validation)
+        loss = self.compute_loss(targets, est_targets, infos=infos)
+        self.loss_backward_and_step(loss, validation=validation)
+        return loss
+
+    def unpack_data(self, data):
+        """ Unpack data given by the DataLoader """
+        if len(data) == 2:
+            inputs, targets = data
+            infos = dict()
+        elif len(data) == 3:
+            inputs, targets, infos = data
+        else:
+            raise ValueError('Expected DataLoader output to have '
+                             '2 or 3 elements. Received '
+                             '{} elements'.format(len(data)))
+        if self.use_cuda:
+            inputs, targets, infos = to_cuda([inputs, targets, infos])
+        return inputs, targets, infos
+
+    def forward_model(self, inputs, validation=False):
+        """ Forwards the inputs through the model """
+        if validation:
+            with torch.no_grad():
+                est_targets = self.model(inputs)
+        else:
+            est_targets = self.model(inputs)
+        return est_targets
+
+    def compute_loss(self, targets, est_targets, infos=None):
+        """ Computes the loss to be backpropred.
+        `infos` is a dictionary that is passed to the loss function.
+        """
+        return self.loss_class.compute(targets, est_targets, infos=infos)
+
+    def loss_backward_and_step(self, loss, validation=False):
+        """ Backpropagation of the loss and gradient descend step."""
+        if not validation and not torch.isnan(loss):
```

Comment:
What happens if loss is NaN? May be do that check in compute_loss. 
How about making the some of these functions private? Should they be accessible outside the class?

Reply:
If loss is NaN, optimizer.step() makes all the weights NaN, and the network is unrecoverable.
The good approach will be to find the reason and prevent it. See [the anomaly detection tool from pytorch](https://pytorch.org/docs/master/autograd.html#anomaly-detection). I need to take time to set it up.

Reply:
Put a Warning  for now.